### PR TITLE
Add magnetic field example

### PR DESCRIPTION
**Pull request by akaszynski**
Adds two new example datasets generated from [emg3d - Minimum working example](https://emsig.xyz/emg3d-gallery/gallery/tutorials/minimum_example.html#compute-the-electric-field). @prisae, your stuff is awesome, and I've been wanting to plot magnetic fields for a month now.

```py
>>> import numpy as np
>>> from pyvista import examples
>>> grid = examples.download_dipole_hfield()
>>> data = (grid['hfield_fx']*5E10)
>>> grid['data'] = data
>>> grid.plot(volume=True, opacity='sigmoid_4')
```

https://user-images.githubusercontent.com/11981631/180624510-609aaf12-b52f-4b9c-b2a9-54b2c97d2464.mp4

### Bonus

- Improve our opacity parameter docstring in ``add_volume``.
- Improve our invalid opacity error.

### Notes

It was entirely not obvious that the scalars had to be between 0 and 255 for this to work. We should either provide more examples or perhaps a warning when the scalars are `np.uint8`, or ideally both. Volumetric plots are too cool to not have this spelled out.

### CODE REVIEW

File: pyvista/examples/downloads.py [hunk 1]
```diff
@@ -4293,3 +4293,92 @@ def download_particles_lethe(load=True):  # pragma: no cover
...
+        to ``False`` and only the filename will be returned.
+
+    Returns
+    -------
+    pyvista.UniformGrid or str
+        DataSet or filename depending on ``load``.
+
+    Examples
+    --------
+    Download the electric field dataset and plot it as a volume.
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:4316]:
```suggestion
    Download the electric field dataset and plot a slice of it in the XY plane.
```

File: pyvista/examples/downloads.py [hunk 2]
```diff
@@ -4293,3 +4293,92 @@ def download_particles_lethe(load=True):  # pragma: no cover
...
+        to ``False`` and only the filename will be returned.
+
+    Returns
+    -------
+    pyvista.UniformGrid or str
+        DataSet or filename depending on ``load``.
+
+    Examples
+    --------
+    Download the magnetic field dataset and plot it as a volume.
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:4360]:
```suggestion
    Download the magnetic field dataset and plot a slice of it in the XY plane.
```

File: pyvista/plotting/plotting.py [hunk 3]
```diff
@@ -2629,12 +2629,24 @@ def add_volume(
...
         opacity : str or numpy.ndarray, optional
-            Opacity mapping for the scalars array.
-            A string can also be specified to map the scalars range to a
-            predefined opacity transfer function (options include: 'linear',
-            'linear_r', 'geom', 'geom_r'). Or you can pass a custom made
+            Opacity mapping for the scalars array.  You can pass a custom made
             transfer function that is an array either ``n_colors`` in length or
-            shorter.
+            shorter, or you can pass a string to select the built in transfer
+            function. Should be one of the following:
```
Comment (akaszynski) [inline pyvista/plotting/plotting.py:2635]:
```suggestion
            array, or you can pass a string to select a built in transfer
            function. It a string, should be one of the following:
```

File: pyvista/examples/downloads.py [hunk 4]
```diff
@@ -4293,3 +4293,92 @@ def download_particles_lethe(load=True):  # pragma: no cover
...
+    ...     lighting=False,
+    ... )
+
+    Plot the magnetic field volumetrically. Here we scale the data by ``5E10``
+    to get the data of interest to lie between 0-255.
+
+    >>> import numpy as np
+    >>> from pyvista import examples
+    >>> grid = examples.download_dipole_hfield()
+    >>> data = (grid['hfield_fx']*5E10)
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:4379]:
```suggestion
    >>> data = grid['hfield_fx']*5E10
```

**Comment by adeak** [comment]
Very nice!

When I run your example block locally, I get (beside the plot)
```
2022-07-24 00:13:18.347 (  11.844s) [        5B703740]vtkOpenGLVolumeLookupTa:84    WARN| vtkOpenGLVolumeOpacityTable (0x56ec100): This OpenGL implementation does not support the required texture size of 1048576, falling back to maximum allowed, 16384.This may cause an incorrect lookup table mapping.
```

```
>>> pv.Report()

--------------------------------------------------------------------------------
  Date: Sun Jul 24 00:14:03 2022 CEST

                OS : Linux
            CPU(s) : 8
           Machine : x86_64
      Architecture : 64bit
               RAM : 15.5 GB
       Environment : Python
        GPU Vendor : Intel
      GPU Renderer : Mesa Intel(R) UHD Graphics 620 (WHL GT2)
       GPU Version : 4.6 (Core Profile) Mesa 20.3.5

  Python 3.9.2 (default, Feb 28 2021, 17:03:44)  [GCC 10.2.1 20210110]

           pyvista : 0.36.dev0
               vtk : 9.1.0
             numpy : 1.23.0
           imageio : 2.19.3
           appdirs : 1.4.4
            scooby : 0.5.6
        matplotlib : 3.5.2
           IPython : 8.4.0
          colorcet : 3.0.0
           cmocean : 2.0
        ipyvtklink : 0.2.2
             scipy : 1.8.1
        itkwidgets : 0.32.1
              tqdm : 4.64.0
            meshio : 5.3.4
        jupyterlab : 3.4.3
         pythreejs : Version unknown
--------------------------------------------------------------------------------
```

File: pyvista/examples/downloads.py [hunk 5]
```diff
@@ -4293,3 +4293,92 @@ def download_particles_lethe(load=True):  # pragma: no cover
...
+    ...     log_scale=True,
+    ...     lighting=False,
+    ... )
+
+    Plot the electric field volumetrically.
+
+    >>> import numpy as np
+    >>> from pyvista import examples
+    >>> grid = examples.download_dipole_efield()
+    >>> data = (grid['efield_fx']*2E10)
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:4334]:
```suggestion
    >>> data = grid['efield_fx']*2E10
```

File: pyvista/examples/downloads.py [hunk 6]
```diff
@@ -4293,3 +4293,92 @@ def download_particles_lethe(load=True):  # pragma: no cover
...
+    >>> grid['data'] = data
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:4335]:
```suggestion
    >>> grid['data'] = grid['efield_fx']*2E10
```

File: pyvista/examples/downloads.py [hunk 7]
```diff
@@ -4293,3 +4293,92 @@ def download_particles_lethe(load=True):  # pragma: no cover
...
+    ... )
+
+    Plot the magnetic field volumetrically. Here we scale the data by ``5E10``
+    to get the data of interest to lie between 0-255.
+
+    >>> import numpy as np
+    >>> from pyvista import examples
+    >>> grid = examples.download_dipole_hfield()
+    >>> data = (grid['hfield_fx']*5E10)
+    >>> grid['data'] = data
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:4380]:
```suggestion
    >>> grid['data'] = grid['hfield_fx']*5E10
```

**Comment by akaszynski** [comment]
> Very nice!
> 
> When I run your example block locally, I get (beside the plot)

I have that occur locally as well. Recommend a follow up to clean this up.

Same with the data being between 0-255. That was really frustrating.

File: pyvista/plotting/tools.py [hunk 8]
```diff
@@ -473,7 +487,15 @@ def opacity_transfer_function(mapping, n_colors, interpolate=True, kind='quadrat
...
         except KeyError:
-            raise KeyError(f'opactiy transfer function ({mapping}) unknown.')
+            tf_str = list(transfer_func.keys())
+            tf_str.sort()
+            allowable = (''.join([f'    * "{item}"\n' for item in tf_str]))[:-1]
+            msg = (
+                f'Invalid opactiy transfer function "{mapping}". Should be one '
+                f'the following:\n\n{allowable}'
+            )
+            raise KeyError(KeyErrorMessage(msg))
```
Comment (adeak) [inline pyvista/plotting/tools.py:497]: Just a drive-by remark for now: `KeyErrorMessage` is pretty hacky (I assume the "bug" is still there because you rarely have newlines in keys, and when you do you would _want_ the repr of the key to be printed).

Since it's an implementation detail that we have a dict, how about we just raise a `ValueError`? With my usual recommendation to do `raise ... from None`:
```suggestion
                f'Invalid opacity transfer function "{mapping}". Should be one of '
                f'the following:\n\n{allowable}'
            )
            raise ValueError(msg) from None
```

**Comment by prisae** [comment]
This is awesome @akaszynski ! I will have to try this with a more complex model, such as the SEG salt model: https://emsig.xyz/emg3d-gallery/gallery/models/SEG-EAGE_3D_salt_model.html

**Comment by prisae** [comment]
How did you store the `*.vtk`-files? (`emg3d/dipole/efield.vtk` and `emg3d/dipole/hfield.vtk`)

File: pyvista/examples/downloads.py [hunk 9]
```diff
@@ -4293,3 +4293,90 @@ def download_particles_lethe(load=True):  # pragma: no cover
 
     """
     return _download_and_read('lethe/result_particles.20000.0000.vtu', load=load)
+
+
+def download_dipole_efield(load=True):  # pragma: no cover
+    """Download the electric field dataset as generated by ``emg3d``.
+
+    For more details see `emg3d Minimum Example <https://emsig.xyz/emg3d-gallery/gallery/tutorials/minimum_example.html#sphx-glr-gallery-tutorials-minimum-example-py>`_.
+
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:4302]:
```suggestion

    These examples were generated from the following `script <https://github.com/pyvista/vtk-data/tree/master/Data/emg3d/dipole`>_

```

File: pyvista/examples/downloads.py [hunk 10]
```diff
@@ -4293,3 +4293,90 @@ def download_particles_lethe(load=True):  # pragma: no cover
...
+    """
+    return _download_and_read('emg3d/dipole/efield.vtk', load=load)
+
+
+def download_dipole_hfield(load=True):  # pragma: no cover
+    """Download the magnetic field as generated by ``emg3d``.
+
+    For more details see `emg3d Minimum Example <https://emsig.xyz/emg3d-gallery/gallery/tutorials/minimum_example.html#sphx-glr-gallery-tutorials-minimum-example-py>`_.
+
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:4345]:
```suggestion

    These examples were generated from the following `script <https://github.com/pyvista/vtk-data/tree/master/Data/emg3d/dipole`>_.

```

**Comment by akaszynski** [comment]
> How did you store the `*.vtk`-files? (`emg3d/dipole/efield.vtk` and `emg3d/dipole/hfield.vtk`)

These examples were generated from the following script:
https://github.com/pyvista/vtk-data/tree/master/Data/emg3d/dipole

If you have a submission, I'd love for them to be stored at [pyvista/vtk-data](https://github.com/pyvista/vtk-data).

**Comment by prisae** [comment]
> > How did you store the `*.vtk`-files? (`emg3d/dipole/efield.vtk` and `emg3d/dipole/hfield.vtk`)
> 
> These examples were generated from the following script: https://github.com/pyvista/vtk-data/tree/master/Data/emg3d/dipole
> 
> If you have a submission, I'd love for them to be stored at [pyvista/vtk-data](https://github.com/pyvista/vtk-data).

Interesting. `emg3d` uses under the hood `discretize` for the meshing (if installed). @banesullivan introduced once a `to_vtk`-mixin for it: https://discretize.simpeg.xyz/en/main/api/generated/discretize.mixins.InterfaceVTK.html . However, this is only implemented for cell-centred properties. As the fields live on their corresponding edges, one would first have to interpolate the fields to cell centres, before exporting:
```
grid.write_vtk(
    'efield',
    models={
        'efield_fx': abs(grid.average_edge_x_to_cell * efield.fx.ravel('F')),
        'efield_fy': abs(grid.average_edge_y_to_cell * efield.fz.ravel('F')),
        'efield_fz': abs(grid.average_edge_z_to_cell * efield.fy.ravel('F')),
    },
)
```

**Comment by akaszynski** [comment]
I'll have to use that in the future, though a direct tie in with pyvista seems possible if you ever wanted to be able to directly plot the fields.

**Comment by prisae** [comment]
That could be cool. Currently I have, consciously, no plotting routines implemented in `emg3d` at all. The available plotting happens through `discretize`.

Besides volume plot would, of course, stream plots be very interesting for electromagnetic fields.

image
(from https://www.digitalearthlab.com/8-marine-csem-visualising-the-fields/)

Also cool: https://www.youtube.com/watch?v=EPxNdESwQP0

File: examples/99-advanced/magnetic-fields.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+Plot Electric and Magnetic Fields
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following example demonstrates how PyVista can plot electric and magnetic
+fields.
+
+The modeling results are courtesy of `prisae <https://github.com/prisae>`_ and
+are from the `emg3d Minimum working example
+<https://emsig.xyz/emg3d-gallery/gallery/tutorials/minimum_example.html#sphx-glr-gallery-tutorials-minimum-example-py>`_.
```
Comment (prisae) [inline examples/99-advanced/magnetic-fields.py:12]: You could shorten this
<https://emsig.xyz/emg3d-gallery/gallery/tutorials/minimum_example.html#sphx-glr-gallery-tutorials-minimum-example-py>`_
to
<https://emsig.xyz/emg3d-gallery/gallery/tutorials/minimum_example.html>`_

File: examples/99-advanced/magnetic-fields.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+###############################################################################
+#
+
+hfield_grid = examples.download_dipole_hfield()
+
+hfield = np.vstack((hfield_grid['hfield_fx'], hfield_grid['hfield_fy'], hfield_grid['hfield_fz'])).T
+
+hfield_grid['hfield'] = hfield
+
+hfield_grid.streamlines(source_center=(c, c, c)).plot()
```
Comment (prisae) [inline examples/99-advanced/magnetic-fields.py:73]: If I try to run this script it results in a NameError:
```
NameError: name 'c' is not defined
```

File: examples/99-advanced/magnetic-fields.py (hunk 11 above)
Comment (akaszynski) [reply examples/99-advanced/magnetic-fields.py:12]:
```suggestion
<https://emsig.xyz/emg3d-gallery/gallery/tutorials/minimum_example.html>`_.
```

**Comment by akaszynski** [comment]
Ended up going with [magpylib](https://magpylib.readthedocs.io/) for now.

image

There were issues with grid alignment with `emg3d`. I can return to it later in a follow-up PR.

**Comment by prisae** [comment]
Hm. Interesting. `emg3d` uses `discretize` for the meshing. So that would mean that everything of SimPEG has also this alignment problem...? But what do you mean with alignment issues? The fields in the different directions live on different edges, so they are interpolated if you want the full tensor. But that is all implemented.

**Comment by akaszynski** [comment]
> Hm. Interesting. `emg3d` uses `discretize` for the meshing. So that would mean that everything of SimPEG has also this alignment problem...? But what do you mean with alignment issues? The fields in the different directions live on different edges, so they are interpolated if you want the full tensor. But that is all implemented.

I'd like to have both. I'll open an issue in `emg3d`.

**Comment by akaszynski** [comment]
@prisae, I think my issue is I'm not implementing it correctly. I'll fix it to:

```py
grid.write_vtk(
    'efield',
    models={
        'efield_fx': abs(grid.average_edge_x_to_cell * efield.fx.ravel('F')),
        'efield_fy': abs(grid.average_edge_y_to_cell * efield.fz.ravel('F')),
        'efield_fz': abs(grid.average_edge_z_to_cell * efield.fy.ravel('F')),
    },
)
```

Docs are down at the moment, see https://github.com/pyvista/pyvista/pull/3058.

**Comment by prisae** [comment]
That looks good @akaszynski . The fields indeed live on the corresponding edges. So if you want to use them together it is the easiest to move them to cell centers first.

File: examples/99-advanced/magnetic-fields.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+pl.show()
+
+
+###############################################################################
+# Plot the Magnet Field Strength
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Finally, let's bring this all together by plotting the magnetic field
+# strength while also plotting the steamlines and the coil.
+
+# Take the norm of the magetic field
```
Comment (tkoyama010) [inline examples/99-advanced/magnetic-fields.py:82]:
```suggestion
# Plot the Magnet Field Strength
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Finally, let's bring this all together by plotting the magnetic field
# strength while also plotting the streamlines and the coil.

# Take the norm of the magnetic field
```

File: tests/examples/test_download_files.py [hunk 14]
```diff
@@ -925,6 +925,15 @@ def test_cad_model_case():
     assert dataset.n_points == 7677
 
 
+def test_coil_magnetic_field():
+    filename = examples.download_coil_magnetic_field(load=False)
+    assert os.path.isfile(filename)
+    assert filename.endswith('vti')
+
+    dataset = examples.download_coil_magnetic_field(load=True)
+    assert isinstance(dataset, pv.UniformGrid)
```
Comment (tkoyama010) [inline tests/examples/test_download_files.py:934]:
```suggestion
    assert isinstance(dataset, pv.UniformGrid)
    assert dataset.n_points == 531441
```

**Review by tkoyama010 — CHANGES_REQUESTED** [review]

**Comment by akaszynski** [comment]
This will have to miss this release, and I'd like to make sure we get `emg3d` in as well here.

**Review by tkoyama010 — DISMISSED** [review]

**Comment by akaszynski** [comment]
Recommending splitting this up into two PRs to get this merged. This PR will incorporate the ``magpylib`` example from [Coil Field Lines](https://magpylib.readthedocs.io/en/latest/examples/examples_30_coil_field_lines.html). 

@prisae, if you can get me a vtk dataset that I can use as a vtk file, I'll be able to generate that streamline example. Please post an issue and I can add it. I'm super busy at the moment and can't investigate this further, but I would love to include another example.

**Comment by akaszynski** [comment]
@tkoyama010, would appreciate another review when you get a chance.

**Review by tkoyama010 — APPROVED** [review]
LGTM
